lib/pbench/agent/utils.py

```diff
@@ -254,7 +254,8 @@ def run_command(args, env=None, name=None, logger=None):
         return output
     except subprocess.CalledProcessError as e:
         message = "%s failed: %s" % (name, e.output)
-        logger.error(message)
+        if logger:
+            logger.error(message)
```

Comment:
I see why we need this test; my question is, why do we let `logger` be `None` instead of assigning it a default logger?

Reply:
I was trying to resolve the error (which probably doesn't happen for real) without a lot of coding work. I fixed another analyzer error in this module and this one was easy to "fix". I don't remember how many callers I found, but one was in a click module where this was called without referencing the logger it already has; I nearly changed that. It's possible we could change the signature here to require providing a logger without breaking anything; this change satisfied the scanner with a minimal footprint.

If you object to touching it without a more comprehensive fix, I can just drop this change; but I don't want to spread out further to "fix" it.

Reply:
Well, I don't like the code having the option of just dropping log/diagnostic output.  (And, I suspect, you might be sympathetic to that viewpoint....)  I _think_ I'd rather see an error/crash than have the output just dropped.

I support your desire to avoid spreading change...but, if `logger` is `None`, can't we just "get" a logger to use?  (I don't recall the details, but getting a logger in general is easy...perhaps I don't care too much about which one we get?)

Reply:
Another consideration is logging in utility functions.  If this is too low-level we may want this returned to the call to deal with as required.

Reply:
But for another PR.  =)

Reply:
> Another consideration is logging in utility functions.

This is a very good point...and it feeds into my growing ambivalence about moving away from the Python logger.  The Python logger is so versatile, and I'm kind of inclined to make it do what we want rather than moving away from it altogether...we just need some way to reconcile our "logging needs" with our "user interface requirements".  (That is, for CLI tools, we want the output to go to `stderr`, not a log somewhere, and we want it to be user-relevant, not "internal diagnostics"...unless, of course, the "user" is a Pbench developer....)

As you say, @portante, that's all for another PR.  For this PR, I guess I would vote for rescinding this change.  (I don't want us left in a state in which we quietly (and, perhaps, unexpectedly) eat diagnostics.)

Reply:
> (That is, for CLI tools, we want the output to go to `stderr`, not a log somewhere, and we want it to be user-relevant, not "internal diagnostics"...unless, of course, the "user" is a Pbench developer....)

Except that in many places the pbench-agent code emits logs to `stdout` or `stderr` AND to the `${pbench_log}` file.

Reply:
Yeah...and, I wonder if that's really what we want (in the general case).  That is, typically, CLI tools send their output to `stdout` or `stderr`, and if the user wants to capture that s/he can redirect it (or use `tee` or `screen` or something).  If we want to produce log output in addition to that, we can certainly do so, but that's more like a developer function than a user function, so I wouldn't expect it to be enabled by default...and, for the most part, I wouldn't expect a single message to be delivered to _both_ places...but, if that's really what we want, the Python logger can be configured to do that sort of thing automagically, without having to set up a special utility function to make it happen.

Reply:
Perhaps, but we don't want to introduce partial behavioral changes.  So we need to (in a separate PR) either move the code base to only `stdout` / `stderr` or ensure we have the correct loggers in place to do both.

Regardless, we need to over-haul the pbench-agent side logging.

Reply:
That will resolve this conversation.

Reply:
Yes, at the expense of yet another merge conflict when I still haven't figured out what went wrong with the last merge conflict. (Sigh.)